hi-ml/src/health_ml/run_ml.py

```diff
@@ -0,0 +1,204 @@
+            logging.info("Setting tags from parent run.")
+            self.set_run_tags_from_parent()
+        # do training
+        with logging_section("Model training"):
+            _, storing_logger = model_train(container=self.container,
+                                            num_nodes=self.experiment_config.num_nodes)
+            self.storing_logger = storing_logger
+
+        RUN_CONTEXT.log(name="Train epochs", value=self.container.num_epochs)
+
+    def is_normal_run_or_crossval_child_0(self) -> bool:
+        """
+        Returns True if the present run is a non-crossvalidation run, or child run 0 of a crossvalidation run.
+        """
+        if self.container.perform_cross_validation:
+            return self.container.cross_validation_split_index == 0
+        return True
+
+    @staticmethod
+    def lightning_data_module_dataloaders(data: LightningDataModule) -> Dict[ModelExecutionMode, Callable]:
+        """
+        Given a lightning data module, return a dictionary of dataloader for each model execution mode.
+
+        :param data: Lightning data module.
+        :return: Data loader for each model execution mode.
+        """
+        return {
+            ModelExecutionMode.TEST: data.test_dataloader,
+            ModelExecutionMode.VAL: data.val_dataloader,
+            ModelExecutionMode.TRAIN: data.train_dataloader
+        }
+
+    def run_inference_for_lightning_models(self, checkpoint_paths: List[Path]) -> None:
+        """
+        Run inference on the test set for all models that are specified via a LightningContainer.
+        :param checkpoint_paths: The path to the checkpoint that should be used for inference.
+        """
+        if len(checkpoint_paths) != 1:
+            raise ValueError(f"This method expects exactly 1 checkpoint for inference, but got {len(checkpoint_paths)}")
+        # lightning_model = self.container.model
+
+        # Run Lightning's built-in test procedure if the `test_step` method has been overridden
+        logging.info("Running inference via the LightningModule.test_step method")
+        # Lightning does not cope with having two calls to .fit or .test in the same script. As a workaround for
+        # now, restrict number of GPUs to 1, meaning that it will not start DDP.
+        self.container.max_num_gpus = 1
+        # Without this, the trainer will think it should still operate in multi-node mode, and wrongly start
+        # searching for Horovod
+        if ENV_OMPI_COMM_WORLD_RANK in os.environ:
+            del os.environ[ENV_OMPI_COMM_WORLD_RANK]
+        # From the training setup, torch still thinks that it should run in a distributed manner,
+        # and would block on some GPU operations. Hence, clean up distributed training.
+        if torch.distributed.is_initialized():  # type: ignore
+            torch.distributed.destroy_process_group()  # type: ignore
+        trainer, _ = create_lightning_trainer(self.container, num_nodes=1)
```

Comment:
it looks to me the callbacks issue is addressed because here create_lightning_trainer will create a container that includes all the callbacks, including those happening on_test_epoch_end - we might want to test this is the case